sawbuck_app/src/views/accept_offer_modal.js

```diff
@@ -0,0 +1,206 @@
+}
+
+// Returns a function which sets id and label keys for the incoming holdings
+const inSetter = state => (id, asset, hasNew) => () => {
+  outSetter(state, 'in')(id, asset)()
+  state.hasNewHolding = hasNew
+}
+
+const countSetter = state => inQuantity => {
+  state.acceptance.count = Math.floor(inQuantity / state.offer.sourceQuantity)
+  state.inQuantity = state.acceptance.count * state.offer.sourceQuantity
+  state.outQuantity = state.acceptance.count * state.offer.targetQuantity
+}
+
+// Returns a function to map holding data to dropdown options
+const optionMapper = (state, key = 'in') => {
+  const setter = key !== 'out' ? inSetter(state) : outSetter(state)
+
+  return holding => ({
+    isSelected: () => state.acceptance[key] === holding.id,
+    text: `${holding.asset} (${holding.label || holding.id})`,
+    onclick: setter(holding.id, holding.asset)
+  })
+}
+
+// Returns a dropdown option which will trigger holding creation
+const newOption = state => ({
+  isSelected: () => !!state.hasNewHolding,
+  text: m('em', 'New Holding'),
+  onclick: inSetter(state)(null, 'New Holding', true)
+})
+
+// Adds a check mark to the appropriate holding option
+const checkMarkMapper = (state, key = 'in') => option => {
+  return _.assign({}, option, {
+    text: [
+      checkMark(option.isSelected()),
+      option.text
+    ]
+  })
+}
+
+// Returns true or false depending on whether or not the form is valid
+const isFormValid = state => !!state.acceptance.count
+
+// Returns a function which will submit a new offer, and holding if applicable
+const submitter = (state, onDone) => () => {
+  return Promise.resolve()
+    .then(() => {
+      if (state.hasNewHolding) {
+        const holdingKeys = ['label', 'description', 'asset']
+        return api.post('holdings', _.pick(state.holding, holdingKeys))
+      }
+    })
+    .then(holding => {
+      const acceptance = {
+        count: state.acceptance.count,
+        source: state.acceptance.out,
+        target: state.acceptance.in
+      }
+      if (holding) acceptance.target = holding.id
+      return api.patch(`offers/${state.offer.id}/accept`, acceptance)
+    })
+    .then(onDone)
+    .then(() => m.route.set('/'))
+}
+
+const getAsset = (id, holdings) => {
+  return holdings.find(holding => holding.id === id).asset
+}
+
+// A versatile modal allowing users to create new offers (and new holdings)
+const AcceptOfferModal = {
+  oninit (vnode) {
+    vnode.state.holding = {}
+    vnode.state.acceptance = {}
+    vnode.state.hasNewHolding = false
+
+    api.get(`offers/${vnode.attrs.offerId}`)
+      .then(offer => {
+        if (offer.error) return console.error(offer.error)
+        vnode.state.offer = offer
+
+        return Promise.all([
+          api.get(`accounts/${api.getPublicKey()}`),
+          api.get(`accounts/${offer.owners[0]}`)
+        ])
+      })
+      .then(([ user, owner ]) => {
+        const inAsset = getAsset(vnode.state.offer.source, owner.holdings)
```

Comment:
Based on feedback from @boydjohnson, is there a condition where the holdings is null?

Reply:
Shouldn't be. Accounts should be sent back with `"holding": []` if they have no holdings.